backend/src/game.ts

```diff
@@ -0,0 +1,197 @@
+	questionText: string;
+	answerTexts: string[];
+	correctAnswers: number[];
+	note?: string;
+	time?: number;
+	points?: number;
+}
+
+export interface Quiz {
+	meta: {
+		title: string;
+		author: string;
+		pointDefault: number;
+		timeDefault: number;
+		note?: string;
+	};
+	questions: QuizQuestion[];
+}
+
+export interface Game {
+	users: Map<UserId, User>;
+	hostId: UserId;
+	activeQuestion: number;
+	quizData: Quiz;
+};
+
+function getUsers(game: Game) {
+	return [...(game.users.keys(), game.hostId)];
+}
+
+function beginQuestion() {
+    return;
+}
+
+// first key is gameId
+const games: Map<GameId, Game> = new Map();
+
+export default function registerGameRoutes(app: Express) {
+	app.post('/games', (req, res) => {
+		try {
+            const quizData: Quiz = req.body;
+			const response = {
+				gameId: code.gen(5, []),
+				hostId: code.gen(8, []),
+			};
+			// associate gameId with data and host
+			const data = {
+				users: new Map(),
+				quizData,
+				hostId: response.hostId,
+				activeQuestion: -1,
+			};
+			games.set(response.gameId, data);
+            console.log(response);
+			res.send(response);
+		} catch (e) {
+			// client upload error
+            console.log(e);
+			res.status(400).send('Invalid JSON file');
+			// Todo: further validation
+		}
+	});
+
+	app.post('/games/:gameId/questions/:index/start', (req, res) => {
+		const gameId: GameId = req.params.gameId;
+		const index = parseInt(req.params.index);
+		const game = games.get(gameId);
+
+		// client-requested game error
+		if (game === undefined) {
+			res.status(404).send({ ok: false , err: `Game ${gameId} not found`});
+			return;
+		}
+
+		// we have no way of validating the host yet
+		// client permission 403 error
+
+		const quiz = game.quizData;
+		// out-of-bounds error
+		if (index >= quiz.questions.length) {
+			res.status(404).send({ ok: false , err: `Question ${index} not found`});
+			return;
+		}
+
+		// start accepting answers for the question index
+		game.activeQuestion = index;
+
+		// show question text and answers on both host and player screens
+		beginQuestion();
+        game.users.set('123', { name: 'bob', answers: [] });
+
+		res.status(200).send({ ok: true });
+	});
+
+	app.post('/games/:gameId/questions/:index/answer', (req, res) => {
+		const gameId: GameId = req.params.gameId;
+		const userId: UserId = req.body.userId;
+		const index = parseInt(req.params.index);
+		const game = games.get(gameId);
+
+		let answer: number;
+        try {
+            answer = parseInt(req.body.answer);
+        } catch (e) {
+            answer = -1;
+        }
+
+		// client-requested game error
+		if (game === undefined) {
+			res.status(404).send({ok: false, err: `Game ${gameId} not found`});
+			return;
+		}
+
+		const quiz = game.quizData;
+
+		// check if question is open
+		if (game.activeQuestion != index) {
+			res.status(400).send({ok: false, err: `Question ${index} is not open for answers`});
+			return;
+		}
```

Comment:
Ugly diff since I had to move some things around. Biggest change is that it now makes sure `answer` was actually a number and that it's a valid answer for the current question.

Since `req.body` has gone through a JSON parse, we don't expect `answer` to be a numeric string, we expect it to be an actual number. If it's a string that means it was actually a string in the JSON which is never valid.

```suggestion
// client-requested game error
		if (game === undefined) {
			res.status(404).send({ ok: false, err: `Game ${gameId} not found` });
			return;
		}

		// check if question is open
		if (game.activeQuestion != index) {
			res
				.status(400)
				.send({ ok: false, err: `Question ${index} is not open for answers` });
			return;
		}

		const answer = req.body.answer;
		if (
			typeof answer != 'number' ||
			answer >= game.quizData.questions[game.activeQuestion].answerTexts.length
		) {
			res
				.status(400)
				.send({ ok: false, err: `Answer index ${answer} is not valid.` });
		}
```